src/Mod/TechDraw/App/DrawViewDimension.cpp

```diff
@@ -574,107 +574,130 @@ std::string  DrawViewDimension::getFormatedValue(int partial)
-        userVal.remove(rxUnits);                        //getUserString(defaultDecimals) without units
-
-        QLocale loc;
-        double userValNum = loc.toDouble(userVal);
-
+               
+        // qUserString is the value + unit with default decimals, so extract the unit
+        // we cannot just use unit.getString() because this would convert '�' to 'deg'
+        QRegExp rxUnits(QString::fromUtf8(" \\D*$")); // space + any non digits at end of string
         int pos = 0;
-        if ((pos = rxUnits.indexIn(qUserString, 0)) != -1)  {
-            qUserStringUnits = rxUnits.cap(0);                //entire capture - non numerics at end of qUserString
+        if ((pos = rxUnits.indexIn(qUserString, 0)) != -1) {
+            qUserStringUnits = rxUnits.cap(0); // entire capture - non numerics at end of qUserString
         }
-        formattedValue = userVal;                            //sensible default
-    #if QT_VERSION >= 0x050000
-        formattedValue = QString::asprintf(Base::Tools::toStdString(formatSpecifier).c_str(),userValNum);
-    #else
-        QString qs2;
-        formattedValue = qs2.sprintf(Base::Tools::toStdString(formatSpecifier).c_str(),userValNum);
-    #endif
 
-        QString repl = userVal;
-        if (useDecimals()) {
-            if (showUnits() || (Type.isValue("Angle")) ||(Type.isValue("Angle3Pt")) ) {
-                repl = qUserString;
-            } else {
-                repl = userVal;
+        // we can have 2 possible results:
+        // - the value in the base unit but without displayed unit
+        // - the value + unit (not necessarily the base unit!)
+        // the user can overwrite the decimal settings, so we must in every case use the formatSpecifier
+        // if useDecimals(), then formatSpecifier = global decimals, otherwise it is %.2f
+        QLocale loc;
+        double userVal;
+        bool checkDecimals = true;
+        if (showUnits() || (Type.isValue("Angle")) || (Type.isValue("Angle3Pt"))) {
+            formattedValue = qUserString; // result value + unit (not necessarily base unit!)
+            // remove unit
+            formattedValue.remove(rxUnits);
+            // to number
+            userVal = loc.toDouble(formattedValue);
+            if (userVal >= 1.0)
+                // we can assure we didn't make an error > 10% via getUserString()
+                checkDecimals = false;
+        }
+        if (checkDecimals){
+            // get value in the base unit with default decimals
+            // for the conversion we use the same method as in DlgUnitsCalculator::valueChanged
+            // at first remove space from unit string
+            QRegExp space(QString::fromUtf8("\\s"));
+            QString UnitString = qUserStringUnits;
+            UnitString.remove(space);
+            // if the unit is a metric lenght unit, nothing has to be converted
+            if ( UnitString == QString::fromUtf8("\xC2\xB5m") || UnitString == QString::fromLatin1("mm")

```

Comment:
This whole if-else block is superfluous. Base::Quantity.getUserString() has an overloaded version that gives you the scale factor for free.

Reply:
I don't understand. I mean, yes, I should replace
`double convertValue = Base::Quantity::parse(QString::fromLatin1("1") + UnitString).getValue();`
by
`asQuantity.getUserString(convertValue, UnitString);`
(also in DlgUnitsCalculator::valueChanged)

But the if-else must stay. Take for example the case that val = 0.050000. The unit system is standard. This gives you a factor of 0.001 but in this case the factor should be 1 since 0.05000 is already in mm. If in contrary the unit system is Imperial civil, then the desired unit is ft and we must convert 0.050000 to 0.000164

What would help is a method to get the base length unit for the currently used unit system.

Reply:
> But the if-else must stay.

Why does it handle then µm, mm, nm, m and km but ignores things like cm, dm, pm, ...? What if in the future the core system will support some exotic units like dam (deka meter)?

You say for ImperialCivil the length is in ft but what about the other imperial systems? What is their expected unit for lengths?

IMO, depending on the value of the enum "unitSystem" you must hard-code the scale factors because only then you can guarantee that the numbers are consistent.
In your example file you have chosen values where each time the quantity uses the same scale factor and thus you won't see the possible inconsistency.

But as soon as you have numbers where different scale factors are used (e.g. one value is given in inch and another value in thou) you get wrong numbers.

Reply:
> Why does it handle then µm, mm, nm, m and km but ignores things like cm, dm, pm, ...? 

I think this is not necessary since getUserString() (precisely UnitsSchemaInternal::schemaTranslate() ) only outputs different units in steps of 1000, even if the base unit of the scheme is cm:
![FreeCAD_8iTh0jPpbl](https://user-images.githubusercontent.com/1828501/100815221-b770ac80-3443-11eb-8e7f-d0bac9e6f233.png)

> You say for ImperialCivil the length is in ft but what about the other imperial systems? What is their expected unit for lengths?

Good point. I'll follow your suggestion and determine this.

Reply:
I just pushed an improved PR where I found a solution that doesn't need the if-else and also works for cases with different units:
![FreeCAD_d_VVWyVcdFNl](https://user-images.githubusercontent.com/1828501/100825843-42a86d00-3459-11eb-9c2f-e2456215cee9.png)
![FreeCAD_d_NkRiFTXwSu](https://user-images.githubusercontent.com/1828501/100825845-43410380-3459-11eb-980c-e82c4f961c12.png)

Reply:
> I think this is not necessary since getUserString() (precisely UnitsSchemaInternal::schemaTranslate() ) only outputs different units in steps of 1000, even if the base unit of the scheme is cm:

But are again relying on implementation details of how it works. This is always bad practice because your code fails as soon as the implementation changes.

> I just pushed an improved PR where I found a solution that doesn't need the if-else and also works for cases with different units:

Yes, this looks good and should reliably avoid inconsistencies.

> Note that asQuantity.getUserString(convertValue, UnitString); cannot be used because it delivers for e.g. val = 0.050 and Standard scheme a conversion factor 0.001 but in this case the factor must be 1. Now I remember why we have what we have in DlgUnitsCalculator::valueChanged() and that's the way it works for all values.

OK.